### PR TITLE
Add a lossy constructor for tinystr

### PR DESCRIPTION
**Pull request by cmyr**
An instance where it is easier for me to write the patch and open a PR than to start a discussion issue.

I want to be able to write,

```rust
const COMMON_SCRIPT: TinyStr<4> = TinyStr::some_constructor(b"Zyyy");
```

but as far as I can tell, the only way to do this currently would require the unsafe `from_bytes_unchecked` constructor. I would like to be able to avoid unsafe, as it introduces additional code review obligations.

I feel like one possible solution to this would be a lossy constructor, that handles invalid input by substituting invalid bytes with a valid replacement.

This PR includes one possible version of this. I'm not at all married to this particular implementation; it is offered as an illustration.

### CODE REVIEW

File: utils/tinystr/src/ascii.rs [hunk 1]
```diff
@@ -22,6 +22,37 @@ impl<const N: usize> TinyAsciiStr<N> {
...
+        let len = if bytes.len() > N { N } else { bytes.len() };
+
+        // Indexing is protected by the len check above
+        #[allow(clippy::indexing_slicing)]
+        while i < len {
+            let b = bytes[i];
+            if b > 0 && b < 0x80 {
+                out[i] = b;
+            } else {
+                out[i] = SPACE;
```
Comment (robertbastian) [inline utils/tinystr/src/ascii.rs:45]: suggestion: traditionally, failed encoding conversions use `?` as the replacement character. "m?me" is more recognizable as an encoding issue than "m me". Another advantage is that it doesn't mess with the naive word count (`.split(' ')`).

Comment (sffc) [reply utils/tinystr/src/ascii.rs:45]: +1 for `?`

Comment (cmyr) [reply utils/tinystr/src/ascii.rs]: Thanks, I agree that's a better choice.

**Review by robertbastian — APPROVED** [review]

**Comment by robertbastian** [comment]
@cmyr I just discovered your use case (https://github.com/googlefonts/fontc/blob/9466281d982d49a2b7b19845e5f13d720c1682d4/fontbe/src/features/properties.rs#L19-L25). You can in fact use the already existing `tinystr!` macro, which will produce a compile-time error for invalid input:

```rust
pub const COMMON_SCRIPT: UnicodeShortName = tinystr!(4, "Zyyy");
```

**Comment by cmyr** [comment]
@robertbastian I did miss that, thanks for the pointer :)

**Pull request merged**